### PR TITLE
2202 clarify min max mos internal orders

### LINKED ISSUE #2202: Clarify min / max MOS on internal orders
## Is your feature request related to a problem? Please describe 👀

There is some confusion regarding the terms `Minimum MOS` and `Maximum MOS` which are parameters on an internal order.
The min MOS is a threshold; if the store has fewer months of stock on hand than this value then a reorder quantity is suggested.
The max MOS is a target number of months of stock on hand that the store should have, and is used in the suggested quantity calculation.

There is also a concern that the introduction of the min MOS value will cause problems for existing mSupply users, who are only used to setting a single max MOS value and would expect that the system suggests a reorder quantity if the months of stock on hand is less than this value.

## Describe the solution you'd like 🎁
Rename the parameters:
- `Minimum MOS` to `Reorder threshold MOS`
- `Maximum MOS` to `Target MOS`

Add another option to the `Reorder threshold MOS` selection - `Not set` which sets the `min_months_of_stock` value to `0` for the requisition. If this option is selected, then the `max_months_of_stock` value is used for the reorder threshold.

It was suggested that setting the reorder threshold to the target MOS value when the target is changed would work around the problem for existing users. This however introduces the problem of handling changes to both values. If the threshold is changed and then the target is changed, should we overwrite the threshold value? Simplest to retain another state for the threshold.

### PR DESCRIPTION
**Pull request by fergie-nz**
Fixes #2202 

# 👩🏻‍💻 What does this PR do? 
- Renames strings for minimum re order quantities and target quantities as outlined in #2202 
- Adds a 'not set' minimum which defaults to the target stock level. 

# 🧪 How has/should this change been tested? 
As a server admin
- [ ] Navigate to Replenishment > Internal orders, and select a draft internal order
Note to demonstrate working values, an internal order with a broad range of stock levels (from < 1 month, to multiple months) should be selected to ensure suggested quantities is being displayed correctly.
- [ ] Set the 'reorder threshold MOS' to 'not set'
- [ ] Suggested quantities should be the same as if 'reorder threshold MOS' was set to the same value as 'Target MOS'.
- [ ] Change the 'reorder threshold MOS' back to 'not set'
- [ ] Change 'Target MOS' to a different value
- [ ] Suggested quantities should be the same as if you had also set the 'reorder threshold MOS' to the same value as 'Target MOS'
- [ ] Change the localisation to Portuguese, Spanish, French
- [ ] Ensure that local strings format well within the internal order UI

## 💌 Any notes for the reviewer?
I had a couple of process questions:

I've added some new translation strings and also changed some strings.However these probably aren't good. Should I leave these out? Or create a new issue? I also wanted to add some new strings to tet.json (tetum?), but didn't know where to start. 
What is the usual process for handling string translations in PRs?

My test results seem to vary in what fails every time I run tests. Is this normal? Am I missing something in my set up? They all seem to be related to "unable to open database file", so I am assuming this is part of my set up.

Also related to the PR:

I'm 90% sure that I've covered all cases where 'minimum threshold MOS' are called in both back end and front end, and have it defaulting to 'target MOS' when it is not set. However I'm not certain that I haven't missed some cases. 
I would love to be confident in the test suite in this case! However as mentioned above, it is inconsistent for me.

### CODE REVIEW

File: client/packages/common/src/intl/locales/en/replenishment.json [hunk 1]
```diff
@@ -77,6 +77,7 @@
   "messages.cant-delete-requisitions": "Can only delete requisitions with a status of 'Draft'",
   "messages.changing-max-mos": "This will change the maximum months of stock target.",
   "messages.changing-min-mos": "This will change the minimum months of stock threshold.",
+  "messages.unassign-min-mos": "This will remove the minimum months of stock threshold. Minimum MOS will default to be the same as maximum MOS.",
```
Comment (fergie-nz) [inline client/packages/common/src/intl/locales/en/replenishment.json:80]: I added a more descriptive prompt in the alert modal.

File: client/packages/common/src/intl/locales/es/replenishment.json [hunk 2]
```diff
@@ -63,6 +63,7 @@
   "messages.deleted-shipments_other": "Se eliminaron {{count}} envíos",
   "messages.changing-max-mos": "Esto cambiará el objetivo máximo de meses de existencias.",
   "messages.changing-min-mos": "Esto cambiará el umbral mínimo de meses de existencias.",
+  "messages.unassign-min-mos": "Esto eliminará el umbral mínimo de meses de existencias. Los meses mínimos de stock serán por defecto los mismos que los meses máximos de stock.",
```
Comment (fergie-nz) [inline client/packages/common/src/intl/locales/es/replenishment.json:66]: This is my best guess at the translation. I wonder if I should flag this as unconfirmed? Or omit it and wait for an expert?

File: client/packages/common/src/intl/locales/fr/replenishment.json [hunk 3]
```diff
@@ -35,9 +35,9 @@
   "label.hide-stock-over-minimum": "Masquer articles supérieurs au minimum",
   "label.location": "Emplacement",
   "label.max": "Max",
-  "label.max-months-of-stock": "MDS Maximum",
```
Comment (fergie-nz) [inline client/packages/common/src/intl/locales/fr/replenishment.json:38]: These are again best guesses at translations. (not reviewed by a native speaker)

File: client/packages/common/src/intl/locales/pt/replenishment.json [hunk 4]
```diff
@@ -31,8 +31,8 @@
   "label.hide-stock-over-minimum": "Ocultar estoque acima do mínimo",
   "label.max": "Máximo",
   "label.min": "Mínimo",
-  "label.max-months-of-stock": "Máximo MDE (Meses De Estoque)",
```
Comment (fergie-nz) [inline client/packages/common/src/intl/locales/pt/replenishment.json:34]: These are again best guesses at translations. (not reviewed by a native speaker)

File: client/packages/common/src/intl/locales/tet/replenishment.json [hunk 5]
```diff
@@ -37,7 +37,7 @@
   "label.draft": "Draf",
   "label.finalised": "Finaliza Ona",
   "label.max": "Maximu",
-  "label.max-months-of-stock": "MOS Maximu",
```
Comment (fergie-nz) [inline client/packages/common/src/intl/locales/tet/replenishment.json:40]: These are again best guesses at translations. (not reviewed by a native speaker)

File: server/service/src/requisition/request_requisition/generate.rs [hunk 6]
```diff
@@ -25,9 +25,15 @@ pub fn generate_suggested_quantity(
     }
     let months_of_stock = available_stock_on_hand as f64 / average_monthly_consumption as f64;
 
-    if max_months_of_stock == 0.0
-        || (min_months_of_stock != 0.0 && months_of_stock > min_months_of_stock)
-    {
+    let default_min_months_of_stock: f64;
```
Comment (fergie-nz) [inline server/service/src/requisition/request_requisition/generate.rs:28]: I think this is the only case where minstock is called in the back end for calculations, and have changed this to default to the target MOS when it is 0.

File: client/packages/common/src/intl/locales/es/replenishment.json (hunk 2 above)
Comment (roxy-dao) [reply client/packages/common/src/intl/locales/es/replenishment.json:66]: You can leave translations out... someone will translate it eventually...

Comment (fergie-nz) [reply client/packages/common/src/intl/locales/es/replenishment.json:66]: Should I remove the translation strings which have been previously translated but are now out of date?
I can imagine that it would be easy to lose track of strings if they remain but are wrong.

For example if I change the english string, and now the french string is out of date. Should I remove the french string so that it is obvious that it needs to be retranslated?

File: client/packages/common/src/intl/locales/en/replenishment.json (hunk 1 above)
Comment (roxy-dao) [inline client/packages/common/src/intl/locales/en/replenishment.json:80]: Can you rephrase all this since we're renaming min and max MOS in UI, I think it makes more sense to change all these messages to target + reorder threshold... and having both min/max and threshold/target makes it sound a bit repetitive...

'This will remove the stock threshold. Target months of stock will now be used as the threshold.' or something similar...

File: server/service/src/requisition/request_requisition/generate.rs [hunk 7]
```diff
@@ -25,9 +25,15 @@ pub fn generate_suggested_quantity(
...
+
+    if min_months_of_stock == 0.0 {
+        default_min_months_of_stock = max_months_of_stock
+    } else {
+        default_min_months_of_stock = min_months_of_stock
+    }
```
Comment (roxy-dao) [inline server/service/src/requisition/request_requisition/generate.rs:34]:
```suggestion
    let default_min_months_of_stock = if min_months_of_stock == 0.0 {
        max_months_of_stock
    } else {
        min_months_of_stock
    };
```

File: client/packages/requisitions/src/RequestRequisition/api/hooks/line/useRequestLines.ts [hunk 8]
```diff
@@ -8,12 +8,14 @@ export const useRequestLines = () => {
...
   const { columns, onChangeSortBy, sortBy } = useRequestColumns();
-  const { lines, minMonthsOfStock } = useRequestFields([
+  const { lines, minMonthsOfStock, maxMonthsOfStock } = useRequestFields([
     'lines',
     'minMonthsOfStock',
+    'maxMonthsOfStock',
   ]);
 
   const sorted = useMemo(() => {
+    const minMonthsOfStockDefined = minMonthsOfStock ?? maxMonthsOfStock;
```
Comment (roxy-dao) [inline client/packages/requisitions/src/RequestRequisition/api/hooks/line/useRequestLines.ts:18]: :nit:
```suggestion
    const threshold = minMonthsOfStock ?? maxMonthsOfStock;
```

**Review by roxy-dao — CHANGES_REQUESTED** [review]
Looks good~ thanks Ferg (: 

- In `RequestRequisition/api.ts -> insert`, could you change min_months_of_stock to 0 so that the option defaults to `Not Set`?

Really wondering if we should change `min MOS` to an optional instead... but not in the scope of this issue since does require some refactoring...

**Comment by mark-prins** [comment]
> I've added some new translation strings and also changed some strings.However these probably aren't good. Should I leave these out? Or create a new issue? I also wanted to add some new strings to tet.json (tetum?), but didn't know where to start.

Add them in please, in English anyway. The other languages can be translated using [weblate](https://translate.msupply.org/) which is integrated with our repo. Translators can use weblate and when ready, commit and push directly to `main`. 

I wouldn't add to tet.json unless you speak tetum - the translations will fallback to english if the key isn't available in the local language.

> What is the usual process for handling string translations in PRs?

We just add whatever is required. Translators can see in weblate which strings require translating. If you are modifying an existing string, I would either ask a native speaker directly ( Richard / Pauline / Alain / Yeison ) and have it done as part of the PR, or simply remove the, now incorrect, translation. 

We should probably have a way of flagging items requiring translation.. but don't do anything currently.

> My test results seem to vary in what fails every time I run tests. Is this normal? Am I missing something in my set up? They all seem to be related to "unable to open database file", so I am assuming this is part of my set up.

Is this cargo tests? If yes, try: ` ulimit -n 1000 && cargo test` 😉 
Or run `yarn test` from the repo root, this runs both client and server tests.

> I'm 90% sure that I've covered all cases where 'minimum threshold MOS' are called in both back end and front end, and have it defaulting to 'target MOS' when it is not set. However I'm not certain that I haven't missed some cases.
I would love to be confident in the test suite in this case! However as mentioned above, it is inconsistent for me.

We also don't have a front end test suite which covers everything and that would be good too.

File: client/packages/requisitions/src/RequestRequisition/DetailView/Toolbar/ToolbarActions.tsx [hunk 9]
```diff
@@ -10,6 +10,7 @@ import {
 import { useRequest, useHideOverStocked } from '../../api';
 
 const months = [1, 2, 3, 4, 5, 6];
+const notSet = { label: 'Not Set', value: 0 };
```
Comment (roxy-dao) [inline client/packages/requisitions/src/RequestRequisition/DetailView/Toolbar/ToolbarActions.tsx:13]:
```suggestion
const notSet = { label: 'Not Set', value: 0 };
```

File: client/packages/requisitions/src/RequestRequisition/DetailView/Toolbar/ToolbarActions.tsx [hunk 10]
```diff
@@ -40,21 +46,37 @@ export const ToolbarActions = () => {
...
-                value: numberOfMonths,
-              }))}
-              onChange={(_, option) =>
-                option &&
-                getMinMOSConfirmation({
-                  onConfirm: () => update({ minMonthsOfStock: option.value }),
-                })
+              value={
+                minMonthsOfStock === 0
+                  ? { label: 'Not Set', value: 0 }
```
Comment (roxy-dao) [inline client/packages/requisitions/src/RequestRequisition/DetailView/Toolbar/ToolbarActions.tsx:51]:
```suggestion
                  ? { t('label.not-set'), value: 0 }
```

File: client/packages/requisitions/src/RequestRequisition/DetailView/Toolbar/ToolbarActions.tsx [hunk 11]
```diff
@@ -40,21 +46,37 @@ export const ToolbarActions = () => {
...
+                  : {
+                      label: t('label.number-months', {
+                        count: minMonthsOfStock,
+                      }),
+                      value: minMonthsOfStock,
+                    }
               }
+              width="150px"
+              options={[
+                notSet,
```
Comment (roxy-dao) [inline client/packages/requisitions/src/RequestRequisition/DetailView/Toolbar/ToolbarActions.tsx:61]:
```suggestion
                {
                  label: t('label.not-set'),
                  value: 0,
                },
```

**Review by roxy-dao — CHANGES_REQUESTED** [review]
Thanks for the changes Ferg~

Sorry, I have a few more requests before... There's another place in `requisition_line/chart/mod.rs` that uses min_MOS for calculations, could you update this?

**Comment by fergie-nz** [comment]
> Thanks for the changes Ferg~
> 
> Sorry, I have a few more requests before... There's another place in `requisition_line/chart/mod.rs` that uses min_MOS for calculations, could you update this?

That's awesome thanks Roxy, thank you for picking that up :)

File: server/service/src/requisition_line/chart/mod.rs [hunk 12]
```diff
@@ -152,8 +152,13 @@ impl SuggestedQuantityCalculation {
             average_monthly_consumption: from.requisition_line_row.average_monthly_consumption
                 as f64,
             stock_on_hand: from.requisition_line_row.available_stock_on_hand as u32,
-            minimum_stock_on_hand: from.requisition_line_row.average_monthly_consumption as f64
-                * from.requisition_row.min_months_of_stock,
+            minimum_stock_on_hand: from.requisition_line_row.average_monthly_consumption as f64 * {
```
Comment (roxy-dao) [inline server/service/src/requisition_line/chart/mod.rs:155]: :nit: for readability...

```        
let minimum_stock_on_hand = if from.requisition_row.min_months_of_stock == 0.0 {
            from.requisition_row.max_months_of_stock as f64
        } else {
            from.requisition_row.min_months_of_stock as f64
        };

....

minimum_stock_on_hand: from.requisition_line_row.average_monthly_consumption as f64
   * minimum_stock_on_hand,
```

**Review by roxy-dao — APPROVED** [review]
Thanks for the changes Ferg (:
